app/src/main/kotlin/com/hedvig/app/LanguageService.kt

```diff
@@ -0,0 +1,60 @@
+package com.hedvig.app
+
+import android.content.Context
+import androidx.annotation.MainThread
+import androidx.appcompat.app.AppCompatDelegate
+import androidx.core.content.edit
+import androidx.core.os.LocaleListCompat
+import androidx.preference.PreferenceManager
+import com.hedvig.android.core.common.preferences.PreferenceKey
+import com.hedvig.android.market.Language
+import com.hedvig.android.market.MarketManager
+
+class LanguageService(
+  private val context: Context,
+  private val marketManager: MarketManager,
+) {
+  private val preferences = PreferenceManager.getDefaultSharedPreferences(context)
+
+  /**
+   * Sets the language, and as a side effect, restarts all running activities.
+   * Only safe to call from the Main Thread.
+   */
+  @MainThread
+  fun setLanguage(language: Language) {
+    AppCompatDelegate.setApplicationLocales(LocaleListCompat.forLanguageTags(language.toString()))
+  }
+
+  fun getLanguage(): Language? {
+    val localeList = AppCompatDelegate.getApplicationLocales()
+    if (localeList.isEmpty) {
+      return null
+    }
+    val locale = localeList[0]!!
+    return Language.from(locale.toLanguageTag())
+  }
+
+  fun performOnLaunchLanguageCheck() {
+    val currentLanguage = AppCompatDelegate.getApplicationLocales()
+    // Always default to English, Sweden, if nothing else is set
+    if (currentLanguage.isEmpty) {
+      setLanguage(Language.EN_SE)
+    }
+  }
```

Comment:
Yes my thought was the following.
Before this is ever called, we *must* have gone through the SplashScreen first, which means that the `performOneTimeLanguageMigration` has been called before, which means that the old language was saved in `AppCompatDelegate.getApplicationLocales()`. Which in turn would mean that there would be no need to check for this again in the MarketingScreen, since it'd be set for sure already.

The only case I imagine this not being the case is that the migration never happens, due to the sharedprefs not having anything saved in there in the first place. But which case could that storage be empty, is there any? And if there is any such chance, maybe this should be happening also in the splash screen and not in the marketing screen which is not guaranteed to show, so exactly as you said merging them into one function.

With that said, on a fresh app install, what happens there? Is the `AppCompatDelegate.getApplicationLocales()` populated with anything, or do we need to set it ourselves? The docs unfortunately don't seem to mention this as far as I can tell 👀